### PR TITLE
Metadata retry

### PR DESCRIPTION
**Pull request by apereverzin**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/BPS-709

### Change description ###
`OcrValidationRetryManager` using blob metadata for storing retry related properties.

`ocr-validation-delay-retry-ms` property renamed to `ocr-validation-delay-retry-sec`

`OcrValidationRetryManager` will be autowired into `OcrValidator` in the next PR. An integration test using `azure-storage` docker image will also be introduced in this new PR.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/services/storage/OcrValidationRetryManagerTest.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+    @BeforeEach
+    void setUp() {
+        ocrValidationRetryManager = new OcrValidationRetryManager(2, 120);
+
+        given(blobClient.getProperties()).willReturn(blobProperties);
+    }
+
+    @Test
+    void isReadyToRetry_should_return_true_if_delay_expiration_not_set() {
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/services/storage/OcrValidationRetryManagerTest.java:44]: missing 'is'?

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/services/storage/OcrValidationRetryManagerTest.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+    @Test
+    void isReadyToRetry_should_return_false_if_delay_expiration_time_not_passed() {
+        // given
+        metadata = Map.of(
+            "ocrValidationRetryCount",
+            "1",
+            "ocrValidationRetryDelayExpirationTime",
+                now(EUROPE_LONDON_ZONE_ID).plusSeconds(10).toString()
+        );
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/services/storage/OcrValidationRetryManagerTest.java:62]: formatting

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/services/storage/OcrValidationRetryManagerTest.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        ocrValidationRetryManager = new OcrValidationRetryManager(2, 120);
+
+        given(blobClient.getProperties()).willReturn(blobProperties);
+    }
+
+    @Test
+    void isReadyToRetry_should_return_true_if_delay_expiration_not_set() {
+        // given
+        metadata = emptyMap();
+        given(blobProperties.getMetadata()).willReturn(metadata);
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/services/storage/OcrValidationRetryManagerTest.java:47]: either put that in setUp or use local variable for `metadata`

File: src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/services/storage/OcrValidationRetryManagerTest.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    private BlobClient blobClient;
+
+    @Mock
+    private BlobProperties blobProperties;
+
+    private Map<String, String> metadata;
+
+    @BeforeEach
+    void setUp() {
+        ocrValidationRetryManager = new OcrValidationRetryManager(2, 120);
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/bulkscanprocessor/services/storage/OcrValidationRetryManagerTest.java:38]: can 2 and 120 be constant that are then used in tests to get relavant times?

**Review by rkondratowicz — COMMENTED** [review]
Can we have an integration test, please?

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/storage/OcrValidationRetryManager.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        Map<String, String> blobMetaData = blobClient.getProperties().getMetadata();
+        String retryDelayExpirationTime = blobMetaData.get(RETRY_DELAY_EXPIRATION_TIME_METADATA_PROPERTY);
+        var zipFilename = blobClient.getBlobName();
+        var containerName = blobClient.getContainerName();
+
+        logger.info(
+            "Checking if retry delay expired on file {} in container {}",
+            zipFilename,
+            containerName
+        );
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/storage/OcrValidationRetryManager.java:47]: dunno, i would only be interested if i'm not ready to retry. just a personal opinion here

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/storage/OcrValidationRetryManager.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    private int getRetryCount(Map<String, String> blobMetaData) {
+        String retryCountStr = blobMetaData.get(RETRY_COUNT_METADATA_PROPERTY);
+
+        if (StringUtils.isBlank(retryCountStr)) {
+            return 0;
+        } else {
+            return Integer.parseInt(retryCountStr);
+        }
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/storage/OcrValidationRetryManager.java:93]: looks like a good candidate for ternary return statement

File: src/main/resources/application.yaml [hunk 7]
```diff
@@ -221,4 +221,4 @@ process-payments:
 use-wrapping-zip: true
 
 ocr-validation-max-retries: 2
-ocr-validation-delay-retry-ms: 120000
+ocr-validation-delay-retry-sec: 120
```
Comment (doncem) [inline src/main/resources/application.yaml:224]: a bit small window. especially when only 2 retries i would expect bigger window and/or even apply some geometric progression to make window longer after each retry

Comment (doncem) [reply src/main/resources/application.yaml:224]: p.s. can think about it and apply in another pr ;)

**Review by doncem — APPROVED** [review]

**Review by mdayican — COMMENTED** [review]
Need to be Thread safe.
All blob operations should be synchronised, need to leased so other thread should not change the value.

**Review by rkondratowicz — COMMENTED** [review]
Can you add integration tests? (with dockerized storage)

**Comment by apereverzin** [comment]
`OcrValidationRetryManager` is not called from the task so it is not feasible at the moment to run an integration test involving processing real file uploaded to the docker image of the blob storage.

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/storage/OcrValidationRetryManager.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        LeaseAcquirer leaseAcquirer,
+        @Value("${ocr-validation-max-retries}") int ocrValidationMaxRetries,
+        @Value("${ocr-validation-delay-retry-sec}") int ocrValidationRetryDelaySec
+    ) {
+        this.leaseAcquirer = leaseAcquirer;
+        this.ocrValidationMaxRetries = ocrValidationMaxRetries;
+        this.ocrValidationRetryDelaySec = ocrValidationRetryDelaySec;
+    }
+
+    public boolean isRetryDelayNotSetOrExpired(BlobClient blobClient) {
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/storage/OcrValidationRetryManager.java:42]: Quite a mouthful.
Does this basically mean the passed argument is ready for retry?

Comment (apereverzin) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/storage/OcrValidationRetryManager.java]: Renamed to `isReadyToRetry`

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/services/storage/OcrValidationRetryManagerTest.java [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+        dockerHost = dockerComposeContainer.getServiceHost("azure-storage", 10000);
+    }
+
+    @AfterAll
+    static void tearDownContainer() {
+        dockerComposeContainer.stop();
+    }
+
+    @Test
+    void should_read_blob_and_save_metadata_in_database_when_zip_contains_metadata_and_pdfs()
```
Comment (mdayican) [inline src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/services/storage/OcrValidationRetryManagerTest.java:75]: I think this test has many tests in it. Better to divide and mock the time.

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/storage/OcrValidationRetryManager.java [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+        this.leaseAcquirer = leaseAcquirer;
+        this.ocrValidationMaxRetries = ocrValidationMaxRetries;
+        this.ocrValidationRetryDelaySec = ocrValidationRetryDelaySec;
+    }
+
+    public boolean isReadyToRetry(BlobClient blobClient) {
+        Map<String, String> blobMetaData = blobClient.getProperties().getMetadata();
+        String retryDelayExpirationTime = blobMetaData.get(RETRY_DELAY_EXPIRATION_TIME_METADATA_PROPERTY);
+
+        final boolean retryDelayExpired = isRetryDelayExpired(retryDelayExpirationTime);
```
Comment (mdayican) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/storage/OcrValidationRetryManager.java:46]: if there is no metadata but blob is locked should we return `true`?

Comment (mdayican) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/storage/OcrValidationRetryManager.java:46]: I think we need to lease first?

Comment (apereverzin) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/storage/OcrValidationRetryManager.java:46]: this method is renamed to `isReadyToRetry` (or maybe even `isReadyToBeProcessed` will be better) and its intent is to check if it is possible to retry, does not matter if the file is locked or not. Retry will be executed by the caller of this method. To check the property in metadata lease is not necessary.

Comment (mdayican) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/storage/OcrValidationRetryManager.java:46]: while this returning true meantime another thread can mark the metadata

File: src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/storage/OcrValidationRetryManager.java [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+    /**
+     * Check if number of retries is exceeding ocrValidationMaxRetries,
+     * if not set new retry delay expiration time, increment retryCount and return true,
+     * otherwise return false which means no more retries are possible.
+     *
+     * @return true if retry is possible and new retry delay expiration time has been set,
+     *         false if no more retries are possible
+     */
+    public boolean setRetryDelayIfPossible(BlobClient blobClient) {
+        Map<String, String> blobMetaData = blobClient.getProperties().getMetadata();
```
Comment (mdayican) [inline src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/storage/OcrValidationRetryManager.java:68]: this should be done after blob leased. 
Meantime another tread can increase the number but thread would continue with old value.

Comment (apereverzin) [reply src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/storage/OcrValidationRetryManager.java:68]: Agree with you here

File: src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/services/storage/OcrValidationRetryManagerTest.java [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+        throws Exception {
+        //Given
+        uploadToBlobStorage(SAMPLE_ZIP_FILE_NAME, zipDir("zipcontents/ok"));
+
+        BlobClient blobClient = testContainer.getBlobClient(SAMPLE_ZIP_FILE_NAME);
+
+        leaseAcquirer.ifAcquiredOrElse(
```
Comment (mdayican) [inline src/integrationTest/java/uk/gov/hmcts/reform/bulkscanprocessor/services/storage/OcrValidationRetryManagerTest.java:85]: this needs to be changed?
now you do not have anything with leaseAcquirer.
you can upload and get lease id. there should be sth `uploadToBlobStorageAndLease` method

**Review by mdayican — APPROVED** [review]

**Pull request merged**